src/libcore/yk_swt.rs

```diff
@@ -7,20 +7,82 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-/// The software trace recorder function.
-/// This is a weak language item, it actually resides in libstd. It has to be weak to allow libcore
-/// to call up to libstd (libstd is not a dependency of libcore).
-extern "Rust" {
-    #[cfg_attr(not(stage0), lang="yk_swt_rec_loc")]
-    fn yk_swt_rec_loc(crate_hash: u64, def_idx: u32, bb: u32);
+/// The result of `yk_swt_stop_tracing_impl()` is an array of this C struct.
+#[repr(C)]
+#[derive(Debug)]
+pub struct MirLoc {
+    /// Unique identifier for the crate.
+    pub crate_hash: u64,
+    /// The definition index.
+    pub def_idx: u32,
+    /// The basic block index.
+    pub bb_idx: u32,
+}
+
+/// Wraps the raw C trace buffer and exposes a more "Rusty" interface to it.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct SWTrace {
+    /// A heap allocated array of `MirLoc` structs, which the consumer must free. Ideally we'd
+    /// have this struct implement `std::ops::Drop` or at least provide a method to do the freeing,
+    /// but we can do neither due to libcore restrictions.
+    buf: *mut MirLoc,
+    /// The number of items in the above array.
+    len: u32,
+}
+
+impl SWTrace {
+    /// Returns the number of MIR locations recorded in the trace.
+    pub fn len(&self) -> u32 {
+        self.len
+    }
+
+    /// Return a pointer to the raw trace buffer. Provided so that the consumer can free it.
+    pub fn buf(&self) -> *mut MirLoc {
+        self.buf
+    }
+
+    /// Returns the location at index `idx` or `None` if the index is out of bounds.
+    pub fn loc<'a>(&'a self, idx: u32) -> Option<&'a MirLoc> {
```

Comment:
I suggest we make this panic if `idx` is out of bounds and return `MirLoc` directly rather than through an `Option`. It makes things quite a bit faster and given we'll be using this a lot, it's worth it.